### PR TITLE
Implement weather API with hourly temperature data and unit testcases

### PR DESCRIPTION
**Pull request by Thrinadh3143**
Please review the weather_app.py and test_weather_app.py both the files and let me know any changes or enhancements required!

### CODE REVIEW

File: weather_app.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import requests
+
+# Data Source Class
+class HourlyTemperature:
+    def __init__(self, temperatures: list):
+        self.temperatures = temperatures
```
Comment (leonstaubach) [inline weather_app.py]: you should also add a field to the dates/timestamps so that each temperature value has an associated point in time for better visualization

File: weather_app.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        params = {
+            'latitude': latitude,
+            'longitude': longitude,
+            'hourly': 'temperature_2m'
+        }
+        response = requests.get(self.api_url, params=params)
+        response.raise_for_status()
+        data = response.json()
+        temperatures = data.get('hourly', {}).get('temperature_2m', [])
+        return HourlyTemperature(temperatures=temperatures)
```
Comment (leonstaubach) [inline weather_app.py]: the response should also return you a list of dates you can utilize as time references (see my comment above)

File: weather_app.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    def get_hourly_temperature(self, latitude: float, longitude: float) -> HourlyTemperature:
+        params = {
+            'latitude': latitude,
+            'longitude': longitude,
+            'hourly': 'temperature_2m'
+        }
+        response = requests.get(self.api_url, params=params)
+        response.raise_for_status()
+        data = response.json()
```
Comment (leonstaubach) [inline weather_app.py]: here you could consider more error handling (erroneous response, bad data types etc.)

File: weather_app.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+    def get_hourly_temperatures(self, latitude: float, longitude: float):
+        hourly_temperature = self.weather_service.get_hourly_temperature(latitude, longitude)
+        return {
+            'temperatures': hourly_temperature.temperatures
+        }
+
+# Factory Function
+def create_weather_service(api_url: str) -> WeatherService:
+    return WeatherService(api_url)
```
Comment (leonstaubach) [inline weather_app.py]: the factory has no conditions, it always just builds one type of service, so it's not really useful. You should add at least one more implementation of your `WeatherService` and then have the factory function build it based on an enumeration parameter

File: weather_app.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
+import requests
+
+# Data Source Class
+class HourlyTemperature:
+    def __init__(self, temperatures: list):
+        self.temperatures = temperatures
+
+# Service Class
+class WeatherService:
```
Comment (leonstaubach) [inline weather_app.py]: This class is supposed to be implemented using abstraction and interfaces (i.e. `AbstractWeatherService` defining the methods, and `ApiWeatherService` & `MockedWeatherService` extending this interface with actual implementations).

**Review by leonstaubach — COMMENTED** [review]
Good coding practices, the one important thing to update is the class structure so that instead of just one `WeatherService` you create a dedicated interface with 2 implementations (see my comment in the code review for it)

File: weather_app.py (hunk 5 above)
Comment (Thrinadh3143) [reply weather_app.py]: I have Updated the code and i believe that you see the updated code, please review it and let me know the changes neeeded

File: weather_app.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+            return HourlyTemperature(temperatures=[], timestamps=[])
+
+# Mocked Weather Service Class for testing
+class MockedWeatherService(AbstractWeatherService):
+    def fetch_hourly_temperature(self, latitude: float, longitude: float) -> HourlyTemperature:
+        # Return some mocked data
+        return HourlyTemperature(
+            temperatures=[20, 21, 19, 18],
+            timestamps=["2024-08-01T00:00:00Z", "2024-08-01T01:00:00Z", "2024-08-01T02:00:00Z", "2024-08-01T03:00:00Z"]
+        )
```
Comment (leonstaubach) [inline weather_app.py:47]: great! Small thing to point out is that you could add "value generator" logic to ensure your `MockedWeatherService` does not always generate the same value. There are different levels of sophistication for this, but this way you can test multiple different cases (random data, static data, edge-cases, etc.)

**Review by leonstaubach — APPROVED** [review]
Your updates look great, you're all set.

I just suggested one update for the mocked service, but it's not required for this assignment.
